jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java

```diff
@@ -613,6 +616,12 @@ public void close() throws IOException
             {
                 // Do an async close
                 channelWrite(content, true, new WriteCompleteCB());
+
+                // If an asynchronous error has happened already we can throw it here.
+                // Otherwise the async call to onWriteComplete must abort if there is a failure as onError may not be called.
+                _abortOnCloseError = true;
+                if (_onError != null)
+                    IO.throwCheckedIOException(_onError);
```

Comment:
I think we should be more explicit that this code is in a known race, even though it will work the same:
```suggestion
                Throwable onError = _onError; // Was an exception reported already
                if (onError != null)
                    IO.throwCheckedIOException(onError);
```

@sbordet I told @lachlan-roberts that you wouldn't like this code... yes it is a race for a general type of async exception, but there is a class of exception that will be thrown during the processing of the close, and it is those that we should throw here so the application can know all was not OK.

Reply:
@gregw @lachlan-roberts IIUC this `if` block is the case where we do an async close, which is actually our educated interpretation of what we should do in case of async I/O, because it's underspecified.
I'm not sure it's worth reporting it with an exception, because we already report it with `onError()` (and if we don't we probably should).

I am worried about the case where we throw from close, but also fail the callback, and we will have one thread receiving the exception trying to write the error page, and to do that modify the channel and output states, and another thread failing the callback also modifying the channel and output states, and calling `onError()`.
I feel that we are just asking for troubles, garbled responses, etc.

If there is an exception in an async close, we call `onError()` and that's enough notifications to the application.

Reply:
The problem is that we don't call `onError()` if there is a failure with the async close.
If there is the following code:
```java
httpOutput.close();
asyncContext.complete();
return;
```

then you can't call `onError()` because `asyncContext.complete()` was called and there was no call to `isReady()` which returned false.

Reply:
Moreover,   we can't rely on calling onError for issues at the end of the stream, because when we have an exception or are at end of stream, then isReady() returns true, so the app will loop and not return from onWritePossible, so we will never be able to call onError even if we wanted to!

Reply:
If we don't call `onError()`, which is the usual path for errors in async I/O, why do we even bother?

The exception may or may not be thrown depending on thread scheduling etc, so it's not reliable behavior, we have a race in our code and we don't provide any reliable benefit to user code, just that sometimes, under certain conditions, if the thread scheduling aligns, maybe we throw.

This is the kind of code that in 6 months we are going to have an issue reported and ask ourselves "who put that racy code in there?"

I don't think it's worth it.

Reply:
Because if we don't throw, then we end up aborting the connection rather than sending a 500

Reply:
@gregw _some_ times we send a 500; some other times we don't.
If it was possible to always send a 500, would be good -- even if it was racy.
But if we can't always send a 500, then I don't think it's worth it, also considering that calling `close()` explicitly is not that common.

Reply:
But we also throw from write or flush, so it's just making it the same as those operations

Reply:
But in `write()` and `flush()` there is no race.